module/zfs/zfs_ioctl.c

```diff
@@ -6039,6 +6060,107 @@ zfs_ioc_diff(zfs_cmd_t *zc)
 	return (error);
 }
 
+static int
+dump_zap(objset_t *os, uint64_t obj, zfs_dump_zap_flags_t flags, zfs_file_t *fp)
+{
+	dmu_buf_t *db;
+	dmu_object_info_t doi;
+	uint64_t count;
+	zap_cursor_t zc;
+	zap_attribute_t za;
+	int err;
+
+	/* ensure zap object */
+	err = dmu_buf_hold(os, obj, 0, FTAG, &db, DMU_READ_NO_PREFETCH);
+	if (err != 0)
+		return (err);
+	dmu_object_info_from_db(db, &doi);
+	dmu_buf_rele(db, FTAG);
+	if (DMU_OT_BYTESWAP(doi.doi_type) != DMU_BSWAP_ZAP)
+		return (SET_ERROR(EINVAL));
+
+	err = zap_count(os, obj, &count);
+	if (err != 0)
+		return (err);
+
+	err = zfs_file_write(fp, &count, sizeof (count), NULL);
+	if (err != 0)
+		return (err);
+
+	for (zap_cursor_init(&zc, os, obj);
+	    (err = zap_cursor_retrieve(&zc, &za)) == 0;
+	    zap_cursor_advance(&zc)) {
+		if (flags & ZDZ_VALUE_ONLY) {
+			err = zfs_file_write(fp, &za.za_first_integer,
+			    sizeof (uint64_t), NULL);
+		} else {
+			zap_pair_record_t zpr;
+			zpr.zpr_value = za.za_first_integer;
+			strcpy(zpr.zpr_key, za.za_name);
+			err = zfs_file_write(fp, &zpr, sizeof (zpr), NULL);
+		}
+
+		if (err != 0)
+			break;
+	}
+	zap_cursor_fini(&zc);
+
+	return ((err == ENOENT) ? 0 : err);
+}
+
+/*
+ * inputs:
+ * zc_name		name of snapshot to enumerate
+ * zc_obj		zap object id
+ * zc_cookie		file descriptor to write path data on
+ * zc_flags		flags for operation (zfs_dump_zap_flags_t)
+ *
+ * outputs:
+ * count followed by zap_pair_record_t's to the file descriptor
+ */
+static int
+zfs_ioc_dump_zap(zfs_cmd_t *zc)
```

Comment:
New ioctls should generally use the new, nvlist-based method.  Could we do that here?

Reply:
Makes sense.